### ISSUE TITLE
🐛 biome lint panics when encountering multi-byte characters

### ISSUE THREAD

**Issue by TatuUlmanen** [biomejs/biome#4181]
### Environment information

```block
CLI:
  Version:                      1.9.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v18.17.1"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "bun/1.1.29"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### What happened?

When running `biome lint` on a JavaScript file containing (only) the expression `a + '€'`, Biome encounters an internal panic. The error message indicates an issue with byte indexing for the Euro symbol (€), which uses more than one byte. Here is the exact error:

```
Biome encountered an unexpected error
Source Location: crates/biome_js_factory/src/utils.rs:34:36
Thread Name: biome::worker_1
Message: byte index 1 is not a char boundary; it is inside '€' (bytes 0..3) of `€`
```

The original code that triggered the error was `return v.toFixed(2).replace('.', ',') + ' €';` which I was able to trim down to just `a + '€'` in a separate file while still producing the same error.

If I change the original code to use template literals instead, no panic happens:

    `${v.toFixed(2).replace('.', ',')} €`

### Expected result

Biome should correctly parse and lint the file without crashing, even when handling non-ASCII characters like the Euro symbol.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by dyc3** [comment]
Can you provide a playground link or a reproduction repo?

**Comment by TatuUlmanen** [comment]
@dyc3 This is easily reproducible by running the following in an empty directory:

```
bun init
bun add --dev --exact @biomejs/biome
echo "a + '€'" > index.ts
bunx biome lint index.ts
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue:

Source Location: crates/biome_js_factory/src/utils.rs:34:36
Thread Name: biome::worker_0
Message: byte index 1 is not a char boundary; it is inside '€' (bytes 0..3) of `€`

index.ts internalError/panic  INTERNAL  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ processing panicked: byte index 1 is not a char boundary; it is inside '€' (bytes 0..3) of `€`

  ⚠ This diagnostic was derived from an internal Biome error. Potential bug, please report it if necessary.


Checked 0 files in 203ms. No fixes applied.
internalError/io ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ No files were processed in the specified paths.
```

The same happens when ran via npm.

**Comment by TatuUlmanen** [comment]
It appears the issue was introduced in a [recent commit](https://github.com/biomejs/biome/commit/cc6d34f88a395a4ac35b0d5c9699fd07c5c3b87a) where the string iteration was changed from using char_indices() to bytes().enumerate(). The change to processing strings as raw bytes caused the error, as bytes().enumerate() does not respect UTF-8 character boundaries. This leads to panics when the code attempts to slice strings containing multi-byte characters, like `€`, at invalid byte offsets.

To handle multi-byte UTF-8 characters safely, the file should be reverted to use char_indices() instead of bytes().enumerate(). This ensures that the string is processed as valid UTF-8 characters, preventing the panic from occurring.

The relevant code change would be:

    let mut iter = unescaped_string.char_indices();

Instead of:

    let mut iter = unescaped_string.bytes().enumerate();

**Comment by Conaclos** [comment]
@TatuUlmanen 

As a temporary workaround you can disable the linter rule `style/useTemplate`.

**Comment by lgarron** [comment]
Thanks for the report @TatuUlmanen and the fix @Conaclos!

I was running into issues with the LSP crashing in VS Code due to an emoji in a string construction that could be converted to a template.

Saw there was a recent fix, and a bump to Biome resolved it! 🥳
